### PR TITLE
Fix delete associated files on rename

### PR DESCRIPTION
**Pull request by h3llrais3r**
When the 'move associated files' flag is not enabled, the 'delete non associated files' flag is enabled, and there are associated files extensions, the associated files are still deleted when renaming. This fixes this behaviour.

Fix for https://github.com/SickRage/SickRage/issues/3306

- [ ] PR is based on the DEVELOP branch
- [ ] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [ ] Read [contribution guide](https://github.com/SickRage/SickRage/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by SickRage2** [comment]
Thanks for the pull request! Before a real human comes by, please make sure your PR has all of the below criteria checked

- [x] Give a description on what the PR is for.
- [x] Make sure your PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review

Please make sure you also read [contribution guide](https://github.com/SickRage/SickRage/blob/master/contributing.md) and followed all of the steps.

Thanks!

**Comment by miigotu** [comment]
I think there is a better solution.
If the file is in the show folder, ALL files are to be kept.
`if some_show_path in associated_file_path or sickbeard.MOVE_ASSOCIATED_FILES and associated_file_path.endswith(tuple(sickbeard.ALLOWED_EXTENSIONS.split(","))):`

**Comment by h3llrais3r** [comment]
I think this is a point of discussion. 😉 
I rather have my library clean and similar.
If I only list 'srt' in my 'associated files extensions' I only want them, also after a rename.
If for some kind of reason other file extensions are still there, I want them removed so I have a clean library. 😉

**Comment by h3llrais3r** [comment]
The rename should not delete files, so you can ignore my previous comment. 😄 
You're right, we should not to check the associated files during rename and just rename everything that matches the filename of the video. 
So how about not even using the `list_associated_files` from `postProcessor`?

**Comment by miigotu** [comment]
You could add an extra param to list_associated_files that is set when listing files from the renamer, that lets us know this is an in-showdir listing, to avoid deleting files from the tv show dir

**Comment by h3llrais3r** [comment]
I also thought of that. Will have a look this evening.

**Comment by h3llrais3r** [comment]
@miigotu Do we keep the same logic for the rename also?
I mean, what do we do with this?https://github.com/SickRage/SickRage/blob/develop/sickbeard/postProcessor.py#L207-L213
Do we keep those 2 excludes?

And do we also apply the rename flag here?
https://github.com/SickRage/SickRage/blob/develop/sickbeard/tv.py#L2454-L2455
I would say yes...

**Comment by h3llrais3r** [comment]
@miigotu Please review.
I added the parameter for rename and also fixed the rename of subtitles in a separate subtitle dir. 😉 

log:
```
2017-03-07 20:47:36 DEBUG    Thread-29 :: Parsed C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 into The Walking Dead - S07E10 [ABD: False] [ANIME: False] [whichReg: [u'standard', u'season_only']]
2017-03-07 20:47:36 DEBUG    Thread-29 :: formatting pattern: %SN - S%0SE%0E - %EN %QN -> The Walking Dead - S07E10 - New Best Friends Unknown
2017-03-07 20:47:36 DEBUG    Thread-29 :: Parsed C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 into The Walking Dead - S07E10 [ABD: False] [ANIME: False] [whichReg: [u'standard', u'season_only']]
2017-03-07 20:47:36 DEBUG    Thread-29 :: formatting pattern: Season %0S -> Season 07
2017-03-07 20:47:36 DEBUG    Thread-29 :: Renaming/moving episode from the base path C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.mkv to C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 - New Best Friends Unknown
2017-03-07 20:47:38 DEBUG    Thread-29 :: Found the following associated files for C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.mkv: [u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.bak', u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.nfo', u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.srt']
2017-03-07 20:47:38 DEBUG    Thread-29 :: Allowing associated files for C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.mkv: [u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.bak', u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.nfo', u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.srt']
2017-03-07 20:48:14 DEBUG    Thread-29 :: Found the following associated files for C:\Temp\subtitles\The Walking Dead - S07E10.mkv: [u'C:\\Temp\\subtitles\\The Walking Dead - S07E10.nl.srt']
2017-03-07 20:48:15 DEBUG    Thread-29 :: Allowing associated files for C:\Temp\subtitles\The Walking Dead - S07E10.mkv: [u'C:\\Temp\\subtitles\\The Walking Dead - S07E10.nl.srt']
2017-03-07 20:48:16 DEBUG    Thread-29 :: Files associated to C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.mkv: [u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.bak', u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.nfo', u'C:\\Temp\\completed\\The Walking Dead\\Season 07\\The Walking Dead - S07E10.srt']
2017-03-07 20:48:16 DEBUG    Thread-29 :: Checking if the path C:\Temp\completed\The Walking Dead\Season 07 already exists
2017-03-07 20:48:16 INFO     Thread-29 :: Renaming file from C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.mkv to C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 - New Best Friends Unknown.mkv
2017-03-07 20:48:16 INFO     Thread-29 :: Trying to clean any empty folders under C:\Temp\completed\The Walking Dead\Season 07
2017-03-07 20:48:16 DEBUG    Thread-29 :: Checking if the path C:\Temp\completed\The Walking Dead\Season 07 already exists
2017-03-07 20:48:16 INFO     Thread-29 :: Renaming file from C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.bak to C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 - New Best Friends Unknown.bak
2017-03-07 20:48:16 INFO     Thread-29 :: Trying to clean any empty folders under C:\Temp\completed\The Walking Dead\Season 07
2017-03-07 20:48:16 DEBUG    Thread-29 :: Checking if the path C:\Temp\completed\The Walking Dead\Season 07 already exists
2017-03-07 20:48:16 INFO     Thread-29 :: Renaming file from C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.nfo to C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 - New Best Friends Unknown.nfo
2017-03-07 20:48:16 INFO     Thread-29 :: Trying to clean any empty folders under C:\Temp\completed\The Walking Dead\Season 07
2017-03-07 20:48:16 DEBUG    Thread-29 :: Checking if the path C:\Temp\completed\The Walking Dead\Season 07 already exists
2017-03-07 20:48:16 INFO     Thread-29 :: Renaming file from C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10.srt to C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 - New Best Friends Unknown.srt
2017-03-07 20:48:16 INFO     Thread-29 :: Trying to clean any empty folders under C:\Temp\completed\The Walking Dead\Season 07
2017-03-07 20:48:17 DEBUG    Thread-29 :: Parsed C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 into The Walking Dead - S07E10 [ABD: False] [ANIME: False] [whichReg: [u'standard', u'season_only']]
2017-03-07 20:48:17 DEBUG    Thread-29 :: formatting pattern: %SN - S%0SE%0E - %EN %QN -> The Walking Dead - S07E10 - New Best Friends Unknown
2017-03-07 20:48:17 DEBUG    Thread-29 :: Checking if the path C:\Temp\subtitles already exists
2017-03-07 20:48:17 INFO     Thread-29 :: Renaming file from C:\Temp\subtitles\The Walking Dead - S07E10.nl.srt to C:\Temp\subtitles\The Walking Dead - S07E10 - New Best Friends Unknown.srt
2017-03-07 20:48:17 INFO     Thread-29 :: Trying to clean any empty folders under C:\Temp\subtitles
2017-03-07 20:48:17 DEBUG    Thread-29 :: Setter sets location to C:\Temp\completed\The Walking Dead\Season 07\The Walking Dead - S07E10 - New Best Friends Unknown.mkv
2017-03-07 20:48:17 DEBUG    Thread-29 :: Transaction with 1 queries executed
```

File: sickbeard/postProcessor.py [hunk 1]
```diff
@@ -214,19 +214,25 @@ def recursive_glob(treeroot, pattern):
 
             # Define associated files (all, allowed and non allowed)
             if ek(os.path.isfile, associated_file_path):
-                if sickbeard.MOVE_ASSOCIATED_FILES and associated_file_path.endswith(tuple(sickbeard.ALLOWED_EXTENSIONS.split(","))):
+                if rename:
+                    # always allow during renaming
```
Comment (sharkykh) [inline sickbeard/postProcessor.py]: @h3llrais3r 
I think this can be done a bit simpler:
```python
    if rename or sickbeard.MOVE_ASSOCIATED_FILES and associated_file_path.endswith(tuple(sickbeard.ALLOWED_EXTENSIONS.split(","))):
        file_path_list_to_allow.append(associated_file_path)
    elif sickbeard.DELETE_NON_ASSOCIATED_FILES and not is_allowed_extension:
        file_path_list_to_delete.append(associated_file_path)
```

Correct me if I'm wrong...

Comment (sharkykh) [reply sickbeard/postProcessor.py]: (I corrected my code)

We're adding/changing less code like that. And it should produce the same result.

Comment (h3llrais3r) [reply sickbeard/postProcessor.py]: I rather keep it separated because it's more clear (it's a matter of taste). 😄 
Additionally, with my code, you don't need to execute the logic for `is_allowed_extension` when we are renaming...

Comment (sharkykh) [reply sickbeard/postProcessor.py]: > Additionally, with my code, you don't need to execute the logic for is_allowed_extension when we are renaming...

Actually,
Python evaluates [statements from left to right](http://stackoverflow.com/questions/752373/is-it-safe-to-rely-on-condition-evaluation-order-in-if-statements) (sorry I couldn't find the Python docs reference),
so when `rename == True`, it will immediately jump inside (doesn't check anything after the `or`) 

------

>I rather keep it separated (it's a matter of taste). 😄 

:+1: miigotu will decide :wink:

Comment (h3llrais3r) [reply sickbeard/postProcessor.py]: `is_allowed_extension` is set before the if statement, so it's always executed in your case 😉

Comment (sharkykh) [reply sickbeard/postProcessor.py]: Sorry I wasn't clear in my review.

My proposal is to change this section:
```python
# Define associated files (all, allowed and non allowed)
if ek(os.path.isfile, associated_file_path):
    if rename:
        # always allow during renaming
        file_path_list_to_allow.append(associated_file_path)
    else:
        # check if allowed or not during post processing
        is_allowed_extension = associated_file_path.endswith(tuple(sickbeard.ALLOWED_EXTENSIONS.split(",")))
        if sickbeard.MOVE_ASSOCIATED_FILES and is_allowed_extension:
            file_path_list_to_allow.append(associated_file_path)
        elif sickbeard.DELETE_NON_ASSOCIATED_FILES and not is_allowed_extension:
            file_path_list_to_delete.append(associated_file_path)
```

to this:
```python
# Define associated files (all, allowed and non allowed)
if ek(os.path.isfile, associated_file_path):
    if rename or sickbeard.MOVE_ASSOCIATED_FILES and associated_file_path.endswith(tuple(sickbeard.ALLOWED_EXTENSIONS.split(","))):
        file_path_list_to_allow.append(associated_file_path)
    elif sickbeard.DELETE_NON_ASSOCIATED_FILES and not is_allowed_extension:
        file_path_list_to_delete.append(associated_file_path)
```

File: sickbeard/postProcessor.py [hunk 2]
```diff
@@ -214,19 +219,21 @@ def recursive_glob(treeroot, pattern):
 
             # Define associated files (all, allowed and non allowed)
             if ek(os.path.isfile, associated_file_path):
-                if sickbeard.MOVE_ASSOCIATED_FILES and associated_file_path.endswith(tuple(sickbeard.ALLOWED_EXTENSIONS.split(","))):
+                # check if allowed or not during post processing
+                is_allowed_extension = associated_file_path.endswith(tuple(sickbeard.ALLOWED_EXTENSIONS.split(",")))
+                if sickbeard.MOVE_ASSOCIATED_FILES and is_allowed_extension:
                     file_path_list_to_allow.append(associated_file_path)
-                elif sickbeard.DELETE_NON_ASSOCIATED_FILES:
+                elif sickbeard.DELETE_NON_ASSOCIATED_FILES and not is_allowed_extension:
```
Comment (miigotu) [inline sickbeard/postProcessor.py]: This idea to only delete files that are not in the allowed extension list doesnt sit well with me.
Either files are in the file_path_list_to_allow (when MOVE_ASSOCIATED_FILES and they are in the allow list) or they are to be deleted when DELETE_NON_ASSOCIATED_FILES.
Once it makes it past the first if and DELETE_NON_ASSOCIATED_FILES then ALL should be deleted, whether they are associated or not.

In other words, all files must be in one of the 2 lists, none can be orphaned out and not deleted or moved.

File: sickbeard/postProcessor.py (hunk 1 above)
Comment (sharkykh) [reply sickbeard/postProcessor.py]: Now it's irrelevant haha

**Review by miigotu — CHANGES_REQUESTED** [review]

**Comment by h3llrais3r** [comment]
Updated.
